File: backend/app/agents/entity_url.py
"""Entity URL helpers (parity with lib/agents/entity-url.ts)."""

from __future__ import annotations

import re
from typing import Any

from app.tools.fallback import build_tool_result

# Sentinel: homepage/pricing scrape skipped because competitor/product URL is unknown.
SKIPPED_INFERRED_URL = "skip-inferred-url"

_PLACEHOLDER_COMPETITOR = frozenset(
    {
        "main competitor",
        "competitor",
        "unknown",
        "n/a",
        "na",
        "none",
        "your competitor",
        "the competitor",
    }
)

_PLACEHOLDER_PRODUCT = frozenset(
    {
        "the product",
        "the current product",
        "our product",
        "your product",
        "product",
        "unknown",
        "n/a",
        "na",
    }
)


async def skipped_scrape_promise() -> dict[str, Any]:
    return build_tool_result(
        data={"url": "", "title": "", "markdown": "", "excerpt": ""},
        status="failed",
        source=SKIPPED_INFERRED_URL,
    )


def is_usable_scrape_page(result: Any) -> bool:
    if isinstance(result, BaseException):
        return False
    v = result
    if not isinstance(v, dict):
        return False
    if v.get("source") == SKIPPED_INFERRED_URL:
        return False
    data = v.get("data") or {}
    md = len((data.get("markdown") or "").strip())
    return bool(data.get("url")) and md > 40


def is_placeholder_competitor(name: str | None) -> bool:
    if not name or not name.strip():
        return True
    return name.lower().strip() in _PLACEHOLDER_COMPETITOR


def is_placeholder_product(name: str | None) -> bool:
    if not name or not name.strip():
        return True
    return name.lower().strip() in _PLACEHOLDER_PRODUCT


def competitor_site_url(ctx: dict[str, Any]) -> str | None:
    explicit = (ctx.get("competitor_url") or "").strip()
    if explicit and re.match(r"^https?://", explicit, re.IGNORECASE):
        return explicit
    if is_placeholder_competitor(ctx.get("competitor")):
        return None
    name = ctx["competitor"].strip()
    slug = name.lower().replace(" ", "")
    if len(slug) < 2 or len(slug) > 40:
        return None
    if not re.fullmatch(r"[a-z0-9]+", slug):
        return None
    return f"https://{slug}.com"


def product_site_url(ctx: dict[str, Any]) -> str | None:
    explicit = (ctx.get("product_url") or "").strip()
    if explicit and re.match(r"^https?://", explicit, re.IGNORECASE):
        return explicit
    if is_placeholder_product(ctx.get("product")):
        return None
    name = ctx["product"].strip()
    words = len(name.split())
    if len(name) > 35 or words > 4:
        return None
    slug = name.lower().replace(" ", "")
    if len(slug) < 2 or len(slug) > 40:
        return None
    if not re.fullmatch(r"[a-z0-9]+", slug):
        return None
    return f"https://{slug}.com"

File: lib/agents/entity-url.ts
import { buildToolResult } from '../tools/fallback';
import type { ScrapedPage, ToolResult } from '../tools/types';
import type { AgentContext } from './types';

/** Sentinel: homepage/pricing scrape skipped because competitor/product URL is unknown. */
export const SKIPPED_INFERRED_URL = 'skip-inferred-url';

export function skippedScrapePromise(): Promise<ToolResult<ScrapedPage>> {
  return Promise.resolve(
    buildToolResult<ScrapedPage>({
      data: { url: '', title: '', markdown: '', excerpt: '' },
      status: 'failed',
      source: SKIPPED_INFERRED_URL,
    }),
  );
}

export function isUsableScrapePage(
  result: PromiseSettledResult<ToolResult<ScrapedPage>>,
): result is PromiseFulfilledResult<ToolResult<ScrapedPage>> {
  if (result.status !== 'fulfilled') return false;
  const v = result.value;
  if (v.source === SKIPPED_INFERRED_URL) return false;
  const md = v.data.markdown?.trim().length ?? 0;
  return !!v.data.url && md > 40;
}

/** Classifier / fallback strings that must never become guessed .com URLs. */
const PLACEHOLDER_COMPETITOR = new Set([
  'main competitor',
  'competitor',
  'unknown',
  'n/a',
  'na',
  'none',
  'your competitor',
  'the competitor',
  'relevant competitors',
  'relevant competitor',
  'top competitors',
  'competitors',
  'peer products',
]);

const PLACEHOLDER_PRODUCT = new Set([
  'the product',
  'the current product',
  'our product',
  'your product',
  'product',
  'unknown',
  'n/a',
  'na',
]);

export function isPlaceholderCompetitor(name: string | undefined | null): boolean {
  if (!name?.trim()) return true;
  return PLACEHOLDER_COMPETITOR.has(name.toLowerCase().trim());
}

export function isPlaceholderProduct(name: string | undefined | null): boolean {
  if (!name?.trim()) return true;
  return PLACEHOLDER_PRODUCT.has(name.toLowerCase().trim());
}

/**
 * Only guess https://brand.com when we have a real competitor name from classification.
 * Otherwise return null — agents should rely on SerpAPI / Reddit / HN only.
 */
export function competitorSiteUrl(ctx: Pick<AgentContext, 'competitor' | 'competitorUrl'>): string | null {
  const explicit = ctx.competitorUrl?.trim();
  if (explicit && /^https?:\/\//i.test(explicit)) return explicit;
  if (isPlaceholderCompetitor(ctx.competitor)) return null;
  const name = ctx.competitor!.trim();
  const slug = name.toLowerCase().replace(/\s+/g, '');
  if (slug.length < 2 || slug.length > 40) return null;
  if (!/^[a-z0-9]+$/.test(slug)) return null;
  return `https://${slug}.com`;
}

/**
 * Guess product homepage only for short, brand-like names (not full sentences).
 */
export function productSiteUrl(ctx: Pick<AgentContext, 'product' | 'productUrl'>): string | null {
  const explicit = ctx.productUrl?.trim();
  if (explicit && /^https?:\/\//i.test(explicit)) return explicit;
  if (isPlaceholderProduct(ctx.product)) return null;
  const name = ctx.product!.trim();
  const words = name.split(/\s+/).length;
  if (name.length > 35 || words > 4) return null;
  const slug = name.toLowerCase().replace(/\s+/g, '');
  if (slug.length < 2 || slug.length > 40) return null;
  if (!/^[a-z0-9]+$/.test(slug)) return null;
  return `https://${slug}.com`;
}
